server/createApp.js

```diff
@@ -4,136 +4,115 @@ var methodOverride = require('method-override');
-			for (var key in loggerOptions.tokens) {
-				if (hasOwnProperty.call(loggerOptions.tokens, key) && typeof loggerOptions.tokens[key] === 'function') {
-					morgan.token(key, loggerOptions.tokens[key]);
-				}
-			}
-		}
-
-		app.use(morgan(keystone.get('logger'), loggerOptions));
+	if (this.get('logger')) {
+		router.use(morgan(this.get('logger'), this.get('logger options')));
 	}
-	// Bind custom logging middleware
-	if (keystone.get('logging middleware')) {
-		app.use(keystone.get('logging middleware'));
+
+	// If the user wants to define their own middleware for logging,
+	// they should be able to
+	if (this.get('logging middleware')) {
+		router.use(this.get('logging middleware'));
 	}
 
+	// We should also allow custom logging middleware to exist in the normal middleware flow
+	router.use(function (req, res, next) {
+		keystone.callHook('pre:logger', req, res, next);
+	});
+
 	// unless the headless option is set (which disables the Admin UI),
 	// bind the Admin UI's Dynamic Router
-	if (!keystone.get('headless')) {
-		app.use(function (req, res, next) {
-			keystone.callHook('pre:admin', req, res, next);
-		});
-		app.use('/' + keystone.get('admin path'), require('../admin/server').createDynamicRouter(keystone));
+	if (!this.get('headless')) {
+		router.use('/' + this.get('admin path'), require('../admin/server').createDynamicRouter(this));
 	}
 
 	// Pre bodyparser middleware
-	if (typeof keystone.get('pre:bodyparser') === 'function') {
-		keystone.get('pre:bodyparser')(app);
+	if (typeof this.get('pre:bodyparser') === 'function') {
+		this.get('pre:bodyparser')(router);
 	}
-	app.use(function (req, res, next) {
+	router.use(function (req, res, next) {
 		keystone.callHook('pre:bodyparser', req, res, next);
 	});
 
-	require('./bindBodyParser')(keystone, app);
-	app.use(methodOverride());
+	require('./bindBodyParser')(this, router);
+	router.use(methodOverride());
 
 	// Set language preferences
-	var languageOptions = keystone.get('language options') || {};
+	var languageOptions = this.get('language options') || {};
 	if (!languageOptions.disable) {
-		app.use(language(keystone));
+		router.use(language(this));
 	}
 
 	// Add 'X-Frame-Options' to response header for ClickJacking protection
-	if (keystone.get('frame guard')) {
-		app.use(require('../lib/security/frameGuard')(keystone));
+	if (this.get('frame guard')) {
+		router.use(require('../lib/security/frameGuard')(this));
 	}
 
 	// Pre route config
-	if (typeof keystone.get('pre:routes') === 'function') {
-		keystone.get('pre:routes')(app);
+	if (typeof this.get('pre:routes') === 'function') {
+		this.get('pre:routes')(router);
 	}
-	app.use(function (req, res, next) {
+	router.use(function (req, res, next) {
 		keystone.callHook('pre:routes', req, res, next);
 	});
 
-	// Configure application routes
-	if (typeof keystone.get('routes') === 'function') {
-		keystone.get('routes')(app);
+	// Configure React routes
+	if (this.get('react routes')) {
```

Comment:
I have the impression you copied an old version of createApp… react routes is gone iirc…